### ISSUE TITLE
Reduce dependencies for shadowing

### ISSUE THREAD

**Issue by velnias75** [derklaro/java-spiget-client#11]
I'm using your project as *version checker* for my Spigot plugins: https://github.com/velnias75/plugin-spiget

For this reason I shadow it with https://github.com/johnrengelman/shadow and try to **minimize** it.
But because `Java8SpigetClient(GsonMapper.INSTANCE)` seens to reference almost all of your project's classes, the **shadow** is quite blown up.

Actually I only use

* `final Version latestVersion = spigetClient.latestResourceVersion().resourceId(spigotId).exec().join();`
* `final Resource resourceDetails = spigetClient.resourceDetails().resourceId(spigotId).exec().join();`

Would it be possible to reduce the class dependencies a bit more?

I'm aware that this would not be an easy task, but it would be very nice.

**Comment by derklaro** [comment]
I personally think that excluding the unused classes from shadowing manually (or using a regex, probably easier) would make more sense in this case rather than making the api less accessible (like removing all request creation methods from the client class and require to pass the client to each request, or making the execute method take the client instance) 🤔

**Comment by velnias75** [comment]
I'll give it a try in the hope the dependencies aren't that complex.

**Comment by velnias75** [comment]
Found a solution …

~~In order to test with excluding classes I tried to set up Junit 5 tests, but it fails:~~

```
java.util.concurrent.CompletionException: java.lang.RuntimeException: No-args constructor for class dev.derklaro.spiget.model.Version does not exist. Register an InstanceCreator with Gson for this type to fix this problem.
	at java.base/java.util.concurrent.CompletableFuture.encodeThrowable(CompletableFuture.java:314)
	at java.base/java.util.concurrent.CompletableFuture.uniApplyNow(CompletableFuture.java:683)
	at java.base/java.util.concurrent.CompletableFuture.uniApplyStage(CompletableFuture.java:658)
	at java.base/java.util.concurrent.CompletableFuture.thenApply(CompletableFuture.java:2164)
	at dev.derklaro.spiget.client.AbstractClient.sendRequest(AbstractClient.java:69)
	at dev.derklaro.spiget.request.resource.LatestResourceVersion.exec(LatestResourceVersion.java:49)
[...]
```

~~I have no plan, how to proceed :-(~~

~~You can find the code including test here: https://github.com/velnias75/plugin-spiget/tree/test-cases~~

**Comment by derklaro** [comment]
After seeing this issue again, I will take a look to move stuff around. That will require a v2 release, but seems worth it. I'm currently a bit busy, so please don't expect that to happen in the next few days (or maybe I grab a coffee and do a late-night refactor) :)

**Comment by derklaro** [comment]
This took way too long from my side to do, but I just pushed a new "v2" branch which leaves the responsibility to provide a client to the developer. It's not quite done (maybe there will be some internal naming changes + the readme needs an update) but for testing purposes it should be enough. If you're interested a jackson-based json mapper was added as well.

Example configuration for Java 11 + GSON:
```java
public class MySpigetStuff {

  private static final SpigetClient CLIENT = new Java11SpigetSpigetClient(SpigetClientConfig.create(GsonMapper.INSTANCE));
  
  public CompletableFuture<Version> getLatestVersion(int resourceId) {
    return LatestResourceVersion.create(CLIENT).resourceId(resourceId).exec();
  }
}
```    

For a dependency use jitpack, I will push the release to central when it's done: https://jitpack.io/#derklaro/java-spiget-client/v2-SNAPSHOT
